compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -7176,10 +7177,51 @@ public void visit(BLangUnaryExpr unaryExpr) {
         }
 
         OperatorKind opKind = unaryExpr.operator;
+
         if (opKind == OperatorKind.ADD || opKind == OperatorKind.SUB) {
             createTypeCastExprForUnaryPlusAndMinus(unaryExpr);
         }
 
+        // Replacing unary expression with numeric literal for finite types
+        if ((opKind == OperatorKind.ADD || opKind == OperatorKind.SUB) && unaryExpr.getBType().tag == TypeTags.FINITE
+                && unaryExpr.expr.getKind() == NodeKind.NUMERIC_LITERAL) {
+            BLangExpression exprInUnary = unaryExpr.expr;
+            BLangNumericLiteral numericLiteralInUnary = (BLangNumericLiteral) exprInUnary;
+            Object objectValueInUnary = numericLiteralInUnary.value;
+            String strValueInUnary = String.valueOf(numericLiteralInUnary.value);
+
+            if (OperatorKind.ADD.equals(unaryExpr.operator)) {
+                strValueInUnary = "+" + strValueInUnary;
+            } else if (OperatorKind.SUB.equals(unaryExpr.operator)) {
+                strValueInUnary = "-" + strValueInUnary;
+            }
+
+            if (objectValueInUnary instanceof Long) {
+                objectValueInUnary = Long.parseLong(strValueInUnary);
+            } else if (objectValueInUnary instanceof Double) {
+                objectValueInUnary = Double.parseDouble(strValueInUnary);
+            } else if (objectValueInUnary instanceof String) {
+                objectValueInUnary = strValueInUnary;
+            }
+
+            BLangNumericLiteral newNumericLiteral = (BLangNumericLiteral)
+                    TreeBuilder.createNumericLiteralExpression();
+            newNumericLiteral.kind = NodeKind.NUMERIC_LITERAL;
+            newNumericLiteral.pos = unaryExpr.pos;
+            newNumericLiteral.value = objectValueInUnary;
+            newNumericLiteral.originalValue = strValueInUnary;
+
+            newNumericLiteral.setDeterminedType(exprInUnary.getBType());
+            newNumericLiteral.setBType(exprInUnary.getBType());
+            newNumericLiteral.expectedType = exprInUnary.getBType();
+
+            newNumericLiteral.typeChecked = unaryExpr.typeChecked;
+            newNumericLiteral.constantPropagated = unaryExpr.constantPropagated;
+            newNumericLiteral.isFiniteContext = true;
+
+            result = rewriteExpr(newNumericLiteral);
+            return;
```

Comment:
`return;` is not needed

Reply:
we are creating a literal and rewriting that instead of a unary expression here to the result  because of the unary change we did in the type checker, so if we don't put a return here, the rewritten unary expression will get re-assigned to result right.

Reply:
This is my mistake. I thought this is the end of the function block.